web3wallet/BestPractices (#1328)
WalletConnect/walletconnect-docs: docs/web3wallet/best_practises.mdx

	groupId="w3w"
	activeOptions={["web","android","ios"]}
>

<PlatformTabItem value="web">
// TODO add how to consume session expiry expiry event
</PlatformTabItem>

<PlatformTabItem value="android">
```kotlin
val walletDelegate = object : Web3Wallet.WalletDelegate {
    override fun onRequestExpired(request: Wallet.Model.ExpiredRequest) {
        //Here this event is triggered when a session request expires - update your UI
    }
    ...other callbacks
}
Web3Wallet.setWalletDelegate(walletDelegate)
```
</PlatformTabItem>

<PlatformTabItem value="ios">
```swift
Web3Wallet.instance.requestExpirationPublisher.sink { _ in
    // let user know that request has expired
}.store(in: &publishers)
```
</PlatformTabItem>

</PlatformTabs>

### Expected Errors

While approving or rejecting a session request the following error might occur:
- Invalid session
    - This error might happen when user approves or rejects a session request on expired session
- Session request expired
    - This error might happen when user approves or rejects a session request that already expires
- Timeout
    - It happens when Relay doesn't acknowledge session settle publish within 10s

## Connection State

> A connection state event is emitted whenever a connection state has changed. Web3Wallet client tracks the Internet connection status as well as the Relay wss connection. The connection state is `true` when both Internet connection and Relay are online, and it is set to `false` otherwise.

Author: Web3Wallet client tracks the connection status.

import Container from './../components/Container'
import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'
import PlatformTabs from '../components/PlatformTabs'
import PlatformTabItem from '../components/PlatformTabItem'

# Best Practices

> The purpose of this guide is to show the best prectices in regards of Web3Wallet client usage. The goal is to provide the best user expirience that just works in every circumstances.

Reviewer: ```suggestion
The purpose of this guide is to show the best practices in regards of Web3Wallet client usage. The goal is to provide the best user experience that just works in every circumstances.
```

## Pairing

A piring is a connection between a wallet and a dapp that has fixed permissions to only allow a dapp to propose a session through it. Dapp can propose infinite number of sessions on one pairing. Wallet must use a pair method from Web3Wallet client to pair with dapp.

<PlatformTabs
	groupId="w3w"
	activeOptions={["web","android","ios"]}
>

<PlatformTabItem value="web">
// TODO add how to pair with web3wallet
</PlatformTabItem>

<PlatformTabItem value="android">
```kotlin
val pairingParams = Wallet.Params.Pair(pairingUri)
Web3Wallet.pair(pairingParams, 
    onSuccess = {
        //Subscribed on the pairing topic successfully. Wallet should await for a session proposal
    },
    onError = { error ->
        //Some error happens while pairing - check Expected errors section
    }
}
```
</PlatformTabItem>

<PlatformTabItem value="ios">
```swift
let uri = WalletConnectURI(string: urlString)

if let uri {
    Task {
        try await Web3Wallet.instance.pair(uri: uri)
    }
}
```
</PlatformTabItem>

</PlatformTabs>

### Pairing State

> A pairing state is a primitive exposed by Web3Wallet client for a wallet to indicate whether it should await a session propsal. The pairing state is `true` when a wallet scans a QR and awaits a session proposal. Once the session proposal is received by the wallet, the pairing state is changed to `false`.

Reviewer: ```suggestion
A pairing state is a primitive exposed by Web3Wallet client for a wallet to indicate whether it should await a session proposal. The pairing state is `true` when a wallet scans a QR and awaits a session proposal. Once the session proposal is received by the wallet, the pairing state is changed to `false`.
```

</PlatformTabItem>

<PlatformTabItem value="android">
Session approve

```kotlin
 Web3Wallet.approveSession(approveProposal,
    onSuccess = {
        //Session approval response was sent successfully - update your UI
    }
    onError = { error ->
        //Error while sending session approval - update your UI
    })
```

Session reject
```kotlin
 Web3Wallet.rejectSession(reject,
    onSuccess = {
        //Session rejection response was sent successfully - update your UI
    },
    onError = { error ->
          //Error while sending session rejection - update your UI
    })
```
</PlatformTabItem>

<PlatformTabItem value="ios">
Session Approve
```swift
do {
    try await Web3Wallet.instance.approve(proposalId: proposal.id, namespaces: sessionNamespaces, sessionProperties: proposal.sessionProperties)
    // Update UI, remove loader
} catch {
    // present error
}
```

Session Reject
```swift
do {
    try await Web3Wallet.instance.reject(proposalId: proposal.id, reason: .userRejected)
    // Update UI, remove loader
} catch {
    // present error
}
```
</PlatformTabItem>

</PlatformTabs>

### Session Proposal Expiry

> A session proposal expiry is 5 mins. It means a given proposal is stored for 5 mins in the SDK storage and user has 5 mins for approval or rejection decision. After that time the below event is emmited and proposal modal should be removed from the app's UI.

Reviewer: ```suggestion
A session proposal expiry is 5 mins. It means a given proposal is stored for 5 mins in the SDK storage and user has 5 mins for approval or rejection decision. After that time the below event is emitted and proposal modal should be removed from the app's UI.
```
